broker/broker.go

```diff
@@ -2,53 +2,305 @@ package broker
+	bindingIDLogKey     = "binding-id"
+	detailsLogKey       = "details"
+	asyncAllowedLogKey  = "acceptsIncomplete"
+	operationDataLogKey = "operation-data-recipe-id"
+)
+
 type Broker struct {
-	Config  *config.Config
-	Catalog *catalog.Catalog
-	Logger  *lager.Logger
+	Compose        compose.Client
+	Config         *config.Config
+	ComposeCatalog *catalog.ComposeCatalog
+	Logger         lager.Logger
+}
+
+type Credentials struct {
+	Host     string `json:"host"`
+	Port     string `json:"port"`
+	Name     string `json:"name"`
+	Username string `json:"username"`
+	Password string `json:"password"`
+	URI      string `json:"uri"`
+	JDBCURI  string `json:"jdbcuri"`
+}
+
+type OperationData struct {
+	RecipeID string `json:"recipe_id"`
+	Type     string `json:"type"`
 }
 
-func New(config *config.Config, catalog *catalog.Catalog, logger *lager.Logger) *Broker {
-	return &Broker{
-		Config:  config,
-		Catalog: catalog,
-		Logger:  logger,
+var composeStatus2State = map[string]brokerapi.LastOperationState{
+	"complete": brokerapi.Succeeded,
+	"running":  brokerapi.InProgress,
+	"waiting":  brokerapi.InProgress,
+}
+
+func New(compose compose.Client, config *config.Config, catalog *catalog.ComposeCatalog, logger lager.Logger) (*Broker, error) {
+	if compose == nil {
+		return nil, fmt.Errorf("broker: composer should not be nil")
 	}
+
+	broker := Broker{
+		Compose:        compose,
+		Config:         config,
+		ComposeCatalog: catalog,
+		Logger:         logger,
+	}
+
+	return &broker, nil
 }
 
 func (b *Broker) Services(context context.Context) []brokerapi.Service {
-	return b.Catalog.Services
+	return b.ComposeCatalog.Catalog.Services
 }
 
 func (b *Broker) Provision(context context.Context, instanceID string, details brokerapi.ProvisionDetails, asyncAllowed bool) (brokerapi.ProvisionedServiceSpec, error) {
-	return brokerapi.ProvisionedServiceSpec{}, fmt.Errorf("%s", "Can't provision an instance")
+	b.Logger.Debug("provision", lager.Data{
+		instanceIDLogKey:   instanceID,
+		detailsLogKey:      details,
+		asyncAllowedLogKey: asyncAllowed,
+	})
+
+	spec := brokerapi.ProvisionedServiceSpec{}
+
+	if !asyncAllowed {
+		return spec, brokerapi.ErrAsyncRequired
+	}
+
+	provisionParameters := ProvisionParameters{}
+	if len(details.RawParameters) > 0 {
+		if err := json.Unmarshal(details.RawParameters, &provisionParameters); err != nil {
+			return spec, err
+		}
+	}
+
+	service, err := b.ComposeCatalog.GetService(details.ServiceID)
+	if err != nil {
+		return spec, err
+	}
+
+	plan, err := service.GetPlan(details.PlanID)
+	if err != nil {
+		return spec, err
+	}
+
+	params := composeapi.DeploymentParams{
+		Name:         instanceID,
+		AccountID:    b.Config.AccountID,
+		Datacenter:   "aws:eu-west-1",
+		DatabaseType: service.Name,
+		Units:        plan.Metadata.Units,
+
+		// User provided provisionParameters
+		Version:    provisionParameters.Version,
+		SSL:        !provisionParameters.DisableSSL,
+		WiredTiger: provisionParameters.WiredTiger,
```

Comment:
Is it OK for tenants to choose Wired Tiger at this stage? Do we know how to bill for it? Should it be a separate plan?